File: solutions/swift/tournament/1/Sources/Tournament/Tournament.swift
final class Tournament {
    struct Stats {
        var win: Int = 0
        var draw: Int = 0
        var lose: Int = 0

        var mp: Int {
            win + draw + lose
        }

        var points: Int {
            win * 3 + draw
        }
    }

    func tally(_ input: String) -> String {
        var games: [String: Stats] = [:]
        let lines = input.split(separator: "\n")
        for line in lines {
            let fragments = line.split(separator: ";")
            guard fragments.count == 3 else {
                continue
            }
            let firstCommand = String(fragments[0])
            var firstCommandScore = games[firstCommand, default: Stats()]
            let secondCommand = String(fragments[1])
            var secondCommandScore = games[secondCommand, default: Stats()]
            defer {
                games[firstCommand] = firstCommandScore
                games[secondCommand] = secondCommandScore
            }

            switch fragments[2] {
            case "win":
                firstCommandScore.win += 1
                secondCommandScore.lose += 1
            case "loss":
                firstCommandScore.lose += 1
                secondCommandScore.win += 1
            case "draw":
                firstCommandScore.draw += 1
                secondCommandScore.draw += 1
            default:
                break
            }
        }

        let sorted = games.sorted {
            if $0.value.points == $1.value.points {
                return $0.key < $1.key
            }
            return $0.value.points > $1.value.points
        }

        var result = "Team                           | MP |  W |  D |  L |  P"
        for game in sorted {
            let stats = game.value
            var normalizedName = game.key
            while normalizedName.count < 31 {
                normalizedName.append(" ")
            }
            result += "\n\(normalizedName)|  \(stats.mp) |  \(stats.win) |  \(stats.draw) |  \(stats.lose) |  \(stats.points)"
        }

        return result
    }
}

File: solutions/go/tournament/1/tournament.go
package tournament

import (
	"bytes"
	"errors"
	"fmt"
	"io"
	"sort"
	"strings"
)

type stats struct {
	name string
	win  int
	draw int
	lose int
}

func newStats(name string) *stats {
	return &stats{name: name}
}

func (s *stats) mp() int {
	return s.win + s.draw + s.lose
}

func (s *stats) points() int {
	return 3*s.win + s.draw
}

// Tally reads from input buffer game stats and output in the writer formatted stats table
func Tally(reader io.Reader, writer io.Writer) error {
	buf := new(bytes.Buffer)
	_, err := buf.ReadFrom(reader)
	if err != nil {
		return err
	}
	input := buf.String()

	resultsMap := make(map[string]*stats)

	lines := strings.Split(input, "\n")
	for _, line := range lines {
		if strings.HasPrefix(line, "#") || len(line) == 0 {
			continue
		}
		fragments := strings.Split(line, ";")
		if len(fragments) != 3 {
			return errors.New("wrong input string format")
		}
		firstCommandName := fragments[0]
		firstStats := resultsMap[firstCommandName]
		if firstStats == nil {
			firstStats = newStats(firstCommandName)
			resultsMap[firstCommandName] = firstStats
		}
		secondCommandName := fragments[1]
		secondStats := resultsMap[secondCommandName]
		if secondStats == nil {
			secondStats = newStats(secondCommandName)
			resultsMap[secondCommandName] = secondStats
		}
		switch fragments[2] {
		case "win":
			firstStats.win++
			secondStats.lose++
		case "loss":
			firstStats.lose++
			secondStats.win++
		case "draw":
			firstStats.draw++
			secondStats.draw++
		default:
			return errors.New("wrong game status string")
		}
	}

	_, err = writer.Write([]byte("Team                           | MP |  W |  D |  L |  P\n"))
	if err != nil {
		return err
	}

	var sorted []*stats
	for _, stats := range resultsMap {
		sorted = append(sorted, stats)
	}

	sort.Slice(sorted, func(i, j int) bool {
		if sorted[i].points() == sorted[j].points() {
			return sorted[i].name < sorted[j].name
		}
		return sorted[i].points() > sorted[j].points()
	})

	for _, stats := range sorted {
		normalizeString := stats.name
		for len(normalizeString) < 30 {
			normalizeString += " "
		}
		_, err := writer.Write([]byte(fmt.Sprintf("%s |  %d |  %d |  %d |  %d |  %d\n",
			normalizeString,
			stats.mp(),
			stats.win,
			stats.draw,
			stats.lose,
			stats.points(),
		)))
		if err != nil {
			return err
		}
	}
	return nil
}
